homeassistant/components/recorder/purge.py

```diff
@@ -101,3 +125,174 @@ def purge_old_data(instance, purge_days: int, repack: bool) -> bool:
+
+    # Check if excluded entity_ids are in database
+    excluded_entity_ids: List[str] = [
+        entity_id
+        for (entity_id,) in session.query(distinct(States.entity_id)).all()
+        if not instance.entity_filter(entity_id)
+    ]
+
+    if len(excluded_entity_ids) != 0:
+        batch_purge_before_states = _purge_filtered_states(
+            session,
+            excluded_entity_ids,
+            batch_purge_before_states,
+            batch_size_hours,
+            subset_size,
+        )
+
+    # Check if excluded event_types are in database
+    excluded_event_types: List[str] = [
+        event_type
+        for (event_type,) in session.query(distinct(Events.event_type)).all()
+        if event_type in instance.exclude_t
+    ]
+
+    if len(excluded_event_types) != 0:
+        batch_purge_before_events = _purge_filtered_events(
+            session,
+            excluded_event_types,
+            batch_purge_before_events,
+            batch_size_hours,
+            subset_size,
+        )
+
+    if batch_purge_before_states < utc_now or batch_purge_before_events < utc_now:
+        _LOGGER.debug("Purging filter hasn't fully completed yet")
+        return False
+
+    return True
+
+
+def _purge_filtered_states(
+    session: Session,
+    excluded_entity_ids: List[str],
+    batch_purge_before: datetime,
+    batch_size_hours: int,
+    subset_size: int,
+) -> datetime:
+    """Handle purging filtered states."""
+
+    query = (
+        session.query(States)
+        .filter(States.entity_id.in_(excluded_entity_ids))
+        .order_by(States.last_updated.asc())
+        .limit(1)
+    )
+    states = execute(query, to_native=True, validate_entity_ids=False)
+    if states:
+        batch_purge_before = states[0].last_updated + timedelta(hours=batch_size_hours)
+
+    _LOGGER.debug(
+        "Purging states before %s that should be filtered",
+        batch_purge_before.strftime(DEBUG_DATETIME_FMT),
+    )
+
+    event_ids: List[int] = [
+        event_id
+        for (event_id,) in session.query(States.event_id)
+        .filter(States.last_updated < batch_purge_before)
+        .filter(States.entity_id.in_(excluded_entity_ids))
+        .all()
+    ]
+
+    if event_ids:
+        deleted_rows_states = (
+            session.query(States)
+            .filter(States.last_updated < batch_purge_before)
+            .filter(States.entity_id.in_(excluded_entity_ids))
+            .delete(synchronize_session=False)
+        )
+        _LOGGER.debug(
+            "Deleted %s states because entity_id is excluded", deleted_rows_states
+        )
+
+        deleted_rows_events: int = 0
+        for i in range(0, len(event_ids), subset_size):
+            event_ids_subset = event_ids[i : i + subset_size]
+            deleted_rows_events += (
+                session.query(Events)
+                .filter(Events.event_type == EVENT_STATE_CHANGED)
+                .filter(Events.time_fired < batch_purge_before)
```

Comment:
```suggestion
                .filter(Events.event_type == EVENT_STATE_CHANGED)
                .filter(Events.time_fired < batch_purge_before)
```

If we have the specific event ids, do we need this?

Reply:
I included those filters for query optimization. AFAIK the `in_` check is really inefficient. It compares each `event_id` with each one in the list. The two checks above are executed beforehand and thus limit which `event_ids` are even checked.

Reply:
Makes sense 👍

Reply:
Maybe a subquery to get the event ids ?

Reply:
I want to limit the number of max comparisons. See https://github.com/home-assistant/core/pull/46678#discussion_r577346531 for the detailed explanation

Reply:
> Maybe a subquery to get the event ids ?

I think a subquery wouldn't work here, since the states need to be deleted first.